reps/2022-04-07-serialization.md

```diff
@@ -0,0 +1,162 @@
+In conculution:
+
+1. Provide pluggable ways for users to implement custom serialization, including:
+   1. Cross-language serialization.
+   1. Out of band serialization and other optimizations.
+2. Unify the current serialization code path to this new pluggable design. Make code cleaner.
+### Should this change be within `ray` or outside?
+It should be a part of the main `ray` project.
+## Stewardship
+### Required Reviewers
+Hao Chen, Qing Wang, Eric Liang, Zhi Lin, Simon Mo
+### Shepherd of the Proposal (should be a senior committer)
+Hao Chen
+## Design and Architecture
+### User-level API
+#### Register Serializer
+Firstly, if a user wants to implement custom serialization, he should register his serializer to Ray.
+Note that a unique string ID should be provided for class identification in cross-language serialization.
+```java
+// In Python
+Ray.register_serializer("ArrowTable", type(arrow_table_obj), ArrowTableSerializer());
+Ray.register_serializer("Protobuf", type(protobuf_obj), ProtobufSerializer());
+
+// In Java
+Ray.registerSerializer("ArrowTable", ArrowTable.class, new ArrowTableSerializer());
+Ray.registerSerializer("Protobuf", Protobuf.class, new ProtobufSerializer());
+```
+Ray will maintain a map from ClassID to Serializer in memory. This relationship won't be persisted, which means you need to register them again when the process is restarted.
+#### Implement Serializer
+Then, he needs to implement this custom serializer, implements the interface shown below:
+Let's take Python as an example. Other languages will be similar.
+```python
+class MySerializer(RaySerializer):
+    def serialize(self, object: MyClass) -> RaySerializationResult:
+        pass
+    def deserialize(self, in_band_data: bytes,
+                    out_of_band_data: List[bytes]) -> MyClass:
+        pass
+
+```
+`serialize`method should return a `SerializedResult`, which contains 2 fields: in-band buffer and out-of-band buffers.
+`inBandBuffer` is nothing else than a byte array. Users can use this field to achieve normal in-band serialization.
+```python
+class RaySerializationResult:
+    def __init__(self):
+        self.in_band_buffer: bytes = None
+        self.out_of_band_buffers: List[RayOutOfBandBUffer] = None
+```
+`OutOfBandBuffer` is for advanced users. They can use this to do out-of-band optimization. Ray will call `OutOfBandBuffer#write` only once when constructing the final buffer sent to the network.
+```python
+class MyOutOfBandBuffer(RayOutOfBandBUffer):
+    def get_size(self) -> int:
+        pass
+    # dest is a byte array with size == get_size()
+    def write_to(self, dest: bytes):
+        pass
+```
+### Final Buffer Protocol
+```
++---------------+----------------+-------------------------------------
+|  ClassIDHash  |  in-band data  |   … oob-buf0 … oob-buf1 … oob-buf2 …
```

Comment:
how about also treating in-band data also as a buffer?

Reply:
Fixed, thanks.

Reply:
After thinking again, I think in-band buffer shouldn't be a no-copy buffer. Since it will cause data-sharing. If users want to do zero-copy just use out-of-band buffers.

Reply:
Both in-band/out-band buffer shouldn't be a no-copy buffer due to data copying. But they can be a readonly buffers. My concern with `bytes` it that in Python, converting to `bytes` usually results in extra copy, but I did know some ways to turn readonly buffers to bytes, so I am ok with `bytes` for in-band data.